### ISSUE TITLE
Passing as a URL

### ISSUE THREAD

**Issue by dcurletti** [pavlovml/match#16]
If I wanted to index an image found at this url http://cdn.freshome.com/wp-content/uploads/2013/06/value-outdoor-home-lighting.jpg, what would the workflow be?  Currently, I get: 

```
match_1          | [2017-01-27 23:13:11,640] ERROR in app: Exception on /add [POST]
match_1          | Traceback (most recent call last):
match_1          |   File "/usr/local/lib/python3.5/site-packages/flask/app.py", line 1982, in wsgi_app
match_1          |     response = self.full_dispatch_request()
match_1          |   File "/usr/local/lib/python3.5/site-packages/flask/app.py", line 1614, in full_dispatch_request
match_1          |     rv = self.handle_user_exception(e)
match_1          |   File "/usr/local/lib/python3.5/site-packages/flask/app.py", line 1517, in handle_user_exception
match_1          |     reraise(exc_type, exc_value, tb)
match_1          |   File "/usr/local/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
match_1          |     raise value
match_1          |   File "/usr/local/lib/python3.5/site-packages/flask/app.py", line 1612, in full_dispatch_request
match_1          |     rv = self.dispatch_request()
match_1          |   File "/usr/local/lib/python3.5/site-packages/flask/app.py", line 1598, in dispatch_request
match_1          |     return self.view_functions[rule.endpoint](**req.view_args)
match_1          |   File "/app/match/server.py", line 69, in add_handler
match_1          |     ses.add_image(path, img, bytestream=bs, metadata=metadata)
match_1          |   File "/usr/local/lib/python3.5/site-packages/image_match/signature_database_base.py", line 202, in add_image
match_1          |     rec = make_record(path, self.gis, self.k, self.N, img=img, bytestream=bytestream, metadata=metadata)
match_1          |   File "/usr/local/lib/python3.5/site-packages/image_match/signature_database_base.py", line 339, in make_record
match_1          |     signature = gis.generate_signature(img, bytestream=bytestream)
match_1          |   File "/usr/local/lib/python3.5/site-packages/image_match/goldberg.py", line 168, in generate_signature
match_1          |     avg_grey = self.compute_mean_level(im_array, x_coords, y_coords, P=self.P)
match_1          |   File "/usr/local/lib/python3.5/site-packages/image_match/goldberg.py", line 402, in compute_mean_level
match_1          |     lower_y_lim:upper_y_lim])  # no smoothing here as in the paper
match_1          | TypeError: slice indices must be integers or None or have an __index__ method
```

**Comment by kern** [comment]
Waiting for ascribe/image-match#57 to be fixed, in the meantime just pushed an commit to downgrade to Python 2.

**Comment by robertofabrizi** [comment]
Can you help me get  where you are? I've started the docker container, seems to be working, it spams stuff like this:

```
[2017-01-30 12:10:56 +0000] [298] [INFO] Worker exiting (pid: 298)
[2017-01-30 12:10:56 +0000] [5] [CRITICAL] WORKER TIMEOUT (pid:299)
[2017-01-30 12:10:56 +0000] [299] [INFO] Worker exiting (pid: 299)
[2017-01-30 12:10:56 +0000] [5] [CRITICAL] WORKER TIMEOUT (pid:300)
[2017-01-30 12:10:57 +0000] [300] [INFO] Worker exiting (pid: 300)
[2017-01-30 12:10:57 +0000] [317] [INFO] Booting worker with pid: 317
[2017-01-30 12:10:57 +0000] [318] [INFO] Booting worker with pid: 318
```

I've created the Elasticsearch endpoint on Amazon, it can be reached just fine. I've opened and tested the docker port that I'm using (-p 0.0.0.0:80:8889), port check returns open and telnet connects. 

Yet when I use a website to post a POST request to:

http://ec2-xxxxxx.amazonaws.com:8889/add

passing the parameter url and filepath, the result is that the request times out, and nothing is written in the logs of the docker app.

I'm not really sure what to look for next, can you point me in the right direction?
Thank you so much
Roberto

**Comment by kern** [comment]
You will likely have to clear the image from your Docker cache using `docker rmi pavlov/match:latest`. Then try running the container again. Let me know if this works!

**Comment by robertofabrizi** [comment]
Thank you for the immediate help kern!
Sadly it still does what it did before. I've deleted all docker containers generated from that image, deleted the image, rerun the command

docker run -e ELASTICSEARCH_URL=xxxx.eu-central-1.es.amazonaws.com -p 0.0.0.0:8889:80 -it pavlov/match

This should have started the app (it does, it keeps logging the same entries above) and redirected the port 80 of the container to port 8889 of the host. To double check again that it's not a "firewall" issue or anything after starting the app I checked that I can telnet to the public ip of the host + 8889, so it looks to me like everything is up and running.

Maybe I misinterpreted how to use it? I've opened a browser and simply did a GET request to the public_ip:8889/ping that from my understanding should have returned me the health of the app, but the request times out :(

**Comment by rhsimplex** [comment]
#17 :smiley:

**Comment by kern** [comment]
@robertofabrizi mind removing the image from your Docker cache and repulling? I've updated the image to Python 3 again thanks to `image-match`'s Python 3 fix, which might resolve your issue. :)

**Comment by robertofabrizi** [comment]
@kern  Definitely better now!
Sadly I still get one issue:

`PUT http://search-xxx.eu-central-1.es.amazonaws.com:9200/images [status:N/A request:20.022s]
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/urllib3/connection.py", line 141, in _new_conn
    (self.host, self.port), self.timeout, **extra_kw)
  File "/usr/local/lib/python3.6/site-packages/urllib3/util/connection.py", line 83, in create_connection
    raise err
  File "/usr/local/lib/python3.6/site-packages/urllib3/util/connection.py", line 73, in create_connection
    sock.connect(sa)
socket.timeout: timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/elasticsearch/connection/http_urllib3.py", line 94, in perform_request
    response = self.pool.urlopen(method, url, body, retries=False, headers=self.headers, **kw)
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 649, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/usr/local/lib/python3.6/site-packages/urllib3/util/retry.py", line 324, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/usr/local/lib/python3.6/site-packages/urllib3/packages/six.py", line 686, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 356, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/local/lib/python3.6/http/client.py", line 1239, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/local/lib/python3.6/http/client.py", line 1285, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.6/http/client.py", line 1234, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.6/http/client.py", line 1026, in _send_output
    self.send(msg)
  File "/usr/local/lib/python3.6/http/client.py", line 964, in send
    self.connect()
  File "/usr/local/lib/python3.6/site-packages/urllib3/connection.py", line 166, in connect
    conn = self._new_conn()
  File "/usr/local/lib/python3.6/site-packages/urllib3/connection.py", line 146, in _new_conn
    (self.host, self.timeout))
urllib3.exceptions.ConnectTimeoutError: (<urllib3.connection.HTTPConnection object at 0x7f02f4c6ff98>, 'Connection to search-xxx.eu-central-1.es.amazonaws.com timed out. (connect timeout=10)')`

At the first line it says: http://search-xxx.eu-central-1.es.amazonaws.com:9200/images

Port 9200? I'm not even sure that said port is opened... 
Not sure if it's helpful, but if i go to the url of the elasticsearch via Chrome, port 80 ofc, it says:

`{
  "name" : "Impulse",
  "cluster_name" : "236433397748:xxx",
  "version" : {
    "number" : "2.3.2",
    "build_hash" : "72aa80333344a4fc849da359c9c58acba6c4d9518",
    "build_timestamp" : "2016-11-14T15:59:50Z",
    "build_snapshot" : false,
    "lucene_version" : "5.5.0"
  },
  "tagline" : "You Know, for Search"
}`

**Comment by kern** [comment]
Try removing `:9200` from your `ELASTICSEARCH_URL`.

**Comment by robertofabrizi** [comment]
@kern  I'm not sure that I follow, I never specified any 9200 port, I just started the container like this:

`docker run -e ELASTICSEARCH_URL=xxxx.eu-central-1.es.amazonaws.com -p 0.0.0.0:8889:80 -it pavlov/match`

I'm not sure where that port number comes from...

**Comment by kern** [comment]
Try specifying `:80` as the port?





On Tue, Feb 14, 2017 at 3:11 PM -0800, "Roberto Fabrizi" <notifications@github.com> wrote:












@kern  I'm not sure that I follow, I never specified any 9200 port, I just started the container like this:


docker run -e ELASTICSEARCH_URL=xxxx.eu-central-1.es.amazonaws.com -p 0.0.0.0:8889:80 -it pavlov/match


I'm not sure where that port number comes from...

**Comment by robertofabrizi** [comment]
@kern You mean the host port 80? Sadly I can't do that, there is another service on it, but I don't see how it'd matter, the match application is using port 80 for all it knows, as that is the container port in use by it, it's simply forwarded to port 8889 of the host machine, which the container has no knowledge of...

**Comment by robertofabrizi** [comment]
@kern Ok now I figured out the port thingy. I started elasticsearch from the AWS own console. That only enables port 80 http, not port 9200/9300 tcp.
I've therefore used the dockerized route to start elasticsearch via docker and then on the same machine start match linking the two containers together just like it says in the readme. 
Sadly I'm back to square one again, the containers are up, the port is open, but callint /ping on it doesn't reply...
